File: backend/src/services/ai-ops/stateMachine/definitions/reactDefinition.ts
/**
 * ReAct 编排流程定义
 *
 * 定义 ReAct 编排流程的状态机配置，包含 8 个核心状态节点 + ErrorHandler 错误状态。
 * 转移规则：线性流程 + RoutingDecision 三路分支（FastPath / IntentDrivenExecution / ReActLoop）
 *
 * 需求: 3.1, 3.4, 3.5, 3.6
 */

import { StateDefinition } from '../types';

export const REACT_DEFINITION_ID = 'react-orchestration';

/**
 * ReAct 编排流程状态定义
 *
 * 状态流程：
 *   IntentParse → KnowledgeRetrieval → RoutingDecision
 *     → FastPath → Response
 *     → IntentDrivenExecution → PostProcessing → Response
 *     → ReActLoop → PostProcessing → Response
 *
 * 错误处理：
 *   IntentParse / KnowledgeRetrieval / IntentDrivenExecution / ReActLoop → ErrorHandler → Response
 */
export const reactDefinition: StateDefinition = {
  id: REACT_DEFINITION_ID,
  name: 'ReAct Orchestration Flow',
  version: '1.0.0',
  states: [
    'intentParse',
    'knowledgeRetrieval',
    'routingDecision',
    'fastPath',
    'intentDrivenExecution',
    'reactLoop',
    'postProcessing',
    'response',
    'errorHandler',
  ],
  initialState: 'intentParse',
  terminalStates: ['response', 'errorHandler'],
  transitions: [
    // Linear flow: IntentParse → KnowledgeRetrieval → RoutingDecision
    { from: 'intentParse', to: 'knowledgeRetrieval' },
    { from: 'knowledgeRetrieval', to: 'routingDecision' },

    // RoutingDecision 3-way branch (outcome matching)
    { from: 'routingDecision', to: 'fastPath', condition: 'fastPath' },
    { from: 'routingDecision', to: 'intentDrivenExecution', condition: 'intentDriven' },
    { from: 'routingDecision', to: 'reactLoop', condition: 'reactLoop' },

    // FastPath → Response
    { from: 'fastPath', to: 'response' },

    // IntentDrivenExecution → PostProcessing
    { from: 'intentDrivenExecution', to: 'postProcessing' },

    // ReActLoop → PostProcessing
    { from: 'reactLoop', to: 'postProcessing' },

    // PostProcessing → Response
    { from: 'postProcessing', to: 'response' },

    // Error transitions
    { from: 'intentParse', to: 'errorHandler', condition: 'error' },
    { from: 'knowledgeRetrieval', to: 'errorHandler', condition: 'error' },
    { from: 'intentDrivenExecution', to: 'errorHandler', condition: 'error' },
    { from: 'reactLoop', to: 'errorHandler', condition: 'error' },

    // ErrorHandler → Response (degraded response)
    { from: 'errorHandler', to: 'response' },
  ],
  errorState: 'errorHandler',
  maxSteps: 20,
};

/**
 * ReAct 编排流程的 StateContext 扩展数据类型
 *
 * 通过 context.get/set 存取，定义各阶段写入的数据键和类型。
 */
export interface ReActStateData {
  // 输入数据
  'message': string;
  'conversationContext': unknown;
  'executionContext': unknown;

  // IntentParse 阶段写入
  'parsedIntent': unknown;
  'intentAnalysis': unknown;

  // KnowledgeRetrieval 阶段写入
  'ragContext': unknown;
  'formattedKnowledge': unknown[];
  'knowledgeReferences': unknown[];

  // RoutingDecision 阶段写入
  'routingPath': 'fastPath' | 'intentDriven' | 'reactLoop';
  'routingConfidence': number;

  // ReActLoop / FastPath / IntentDrivenExecution 阶段写入
  'steps': unknown[];
  'finalAnswer': string;
  'iterations': number;

  // PostProcessing 阶段写入
  'validationResult': unknown;
  'reflectionResult': unknown;

  // Response 阶段写入
  'result': unknown;
}

File: opsevo-python/src/opsevo/services/state_machine/definitions/react_definition.py
"""ReAct loop flow definition."""

from ..engine import FlowDefinition, StateDefinition


def create_react_flow() -> FlowDefinition:
    return FlowDefinition(
        name="react_loop",
        initial_state="routing_decision",
        terminal_states={"response", "error"},
        states={
            "routing_decision": StateDefinition(
                name="routing_decision",
                handler="react_routing_decision",
                transitions={"fast_path": "fast_path", "intent_driven": "intent_parse", "react": "knowledge_retrieval"},
            ),
            "fast_path": StateDefinition(
                name="fast_path",
                handler="react_fast_path",
                transitions={"success": "response", "fallback": "knowledge_retrieval", "failure": "error"},
            ),
            "intent_parse": StateDefinition(
                name="intent_parse",
                handler="react_intent_parse",
                transitions={"success": "intent_driven_execution", "failure": "knowledge_retrieval"},
            ),
            "intent_driven_execution": StateDefinition(
                name="intent_driven_execution",
                handler="react_intent_driven_execution",
                transitions={"success": "post_processing", "failure": "knowledge_retrieval"},
                timeout_s=120.0,
            ),
            "knowledge_retrieval": StateDefinition(
                name="knowledge_retrieval",
                handler="react_knowledge_retrieval",
                transitions={"success": "react_loop", "failure": "react_loop"},
            ),
            "react_loop": StateDefinition(
                name="react_loop",
                handler="react_react_loop",
                transitions={"success": "post_processing", "failure": "error"},
                timeout_s=180.0,
            ),
            "post_processing": StateDefinition(
                name="post_processing",
                handler="react_post_processing",
                transitions={"success": "response", "failure": "response"},
            ),
            "response": StateDefinition(name="response", handler="react_response", transitions={}),
            "error": StateDefinition(name="error", handler="react_error_response", transitions={}),
        },
    )
